### ISSUE TITLE
Luigi functionality migration into fragalysis-api

### ISSUE THREAD

**Issue by JenkeScheen** [xchem/fragalysis-api#103]
Hi all,

### Introduction
This thread aims to create an initial mapping of the work to be done in light of recent discussions about troubleshooting current pipeline issues at diamond. Beside fixing congestion in the pipeline and resolving edge-cases, this project will also add some flexibility in fragalysis such that user requests (e.g. from the ASAP comp chem core team) can be implemented.

The intention is that this thread is living, i.e., discussions held here will build a more complete overview of the required work. In this initial post I will map a short overview of the first steps as I understand them, as well as two nice-to-have implementations as requested by ASAP's comp chem core team. Where exactly this thread should be is still open for discussion but I wanted something out this week so I decided to post it here in `fragalysis-api`; we can move the discussion at any point.

--------------
### Initial mapping
The [`luigi` pipeline](https://github.com/xchem/pipeline) that feeds into `fragalysis-api` has issues that result in tasks failing without clear reason. Because of this @Waztom has suggested that some critical functions be migrated into `fragalysis-api` instead. The classes are: `AlignTarget`, `AlignTargetOBO`, `AlignTargetToReference`, `BatchAlignTargets` and `DecideAlignTargets`.  Of these, the parent class is `BatchAlignTargets()` [which is called in the master `luigi`script](https://github.com/xchem/pipeline/blob/master/start_pipeline.py#L58). Although these classes are not fully documented their function is reasonably clear: they involve some decision making but ultimately execute some `sh` scripts that can be migrated to `fragalysis-api` quite easily. From what I understand a reasonable place for this code could be [`fragalysis-api.xcimporter.align`](https://github.com/xchem/fragalysis-api/blob/master/fragalysis_api/xcimporter/align.py).

- [ ] for this migration, a small collection of typical input files for `BatchAlignTargets is needed for testing. Some failing edge-cases would be super helpful here too.

There are 38 occurrences of improper exception handling in the `luigi` codebase which are likely the root cause for the observed unexplained failing tasks. For example, in [`AlignTarget.run()` if there is an issue with the input object the code fails without any type of logging](https://github.com/xchem/pipeline/blob/master/luigi_classes/run_fragalysis.py#L100-L115).  I would suggest we collect a list of functions where this is a high-priority issue and implement proper exception handling to those initially.

### Requests from ASAP team
The comp chem core of ASAP is ramping up its usage of `fragalysis-api` and is requesting two functionalities:
1. (B Kaminow) PDB headers in the `_bound.pdb `and `_apo.pdb` files are missing. From previous discussion we know some of the crystallographic/symmetry stuff involves a bit more work to include, but at least **having the SEQRES header would be quite helpful for building in missing loops/sidechains in an automated way**. A good first step would be to include the removed PDB headers in a separate file in the download, which is already mentioned in the `_bound.pdb` and `_apo.pdb` files, eg:

> $ head -n 1 Mpro-P2007_0A_bound.pdb
> REMARK 777 Header has been moved to Mpro-P2007_0A/Mpro-P2007_0A_header.pdb
      
This should be fairly easy to trace and fix but this message is not present in the codebases of [fragalysis-api](https://github.com/xchem/fragalysis-api), [the `luigi` pipeline](https://github.com/xchem/pipeline) or [XChemExplorer](https://github.com/xchem/XChemExplorer). 
- [ ] what other codebases could this be in?

2. (A Payne) the sdf files of the ligands don't have proper chemical states (bond orders) which currently forces us to load ligand files as PDB and use OpenEye code to guess bond orders. This workaround works well but there may be edge-cases that could be impactful later on in our workflows. 
- [ ] what code handles the generation of SDF files for the `fragalysis-api` downloads?

The above two issues (especially the first one) are high-priority for the ASAP team and would ideally be resolved in parallel with the `luigi` function migration, if possible.

----------------

With that we can start discussing first steps. I will be visiting Diamond Oct 17-21 to get a better overview of all the processes involved after which we can start work on the above points. There are also plans in the works to pull in more developers but details for this are still being worked out.
